crates/sui-types/src/effects/mod.rs

```diff
@@ -190,13 +190,23 @@ impl TransactionEffects {
     }
 }
 
+pub enum InputSharedObjectKind {
+    Mutate,
+    ReadOnly,
+}
+
 #[enum_dispatch]
 pub trait TransactionEffectsAPI {
     fn status(&self) -> &ExecutionStatus;
     fn into_status(self) -> ExecutionStatus;
     fn executed_epoch(&self) -> EpochId;
     fn modified_at_versions(&self) -> &[(ObjectID, SequenceNumber)];
-    fn shared_objects(&self) -> &[ObjectRef];
+    /// Returns the list of shared objects used in the input, with full object reference
+    /// and use kind. This is needed in effects because in transaction we only have object ID
+    /// for shared objects. Their version and digest can only be figured out after sequencing.
+    /// Also provides the use kind to indicate whether the object was mutated or read-only.
+    /// Down the road it could also indicate use-of-deleted.
+    fn input_shared_objects(&self) -> Vec<(ObjectRef, InputSharedObjectKind)>;
```

Comment:
Is the purpose of this function so that effects v2 does not need to store which objects are shared as a separate field, and if so how will we distinguish the object type (shared vs owned) in v2?

Reply:
Correct.
We do store the new owner in effects, and since we don't allow object promotion (owned->shared), if an object exists previously and current owner is shared, it must be mutated shared object.